x-pack/plugins/file_upload/server/telemetry/telemetry.ts

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { Server } from 'hapi';
+import { callWithInternalUserFactory } from '../client/call_with_internal_user_factory';
+
+export const TELEMETRY_DOC_ID = 'file-upload-telemetry';
+
+export interface Telemetry {
+  file_upload: {
+    index_creation_count: number;
+  };
+}
+
+export interface TelemetrySavedObject {
+  attributes: Telemetry;
+}
+
+export function createTelemetry(count: number = 0): Telemetry {
+  return {
+    file_upload: {
+      index_creation_count: count,
+    },
+  };
+}
+
+export function storeTelemetry(server: Server, fileUploadTelemetry: Telemetry): void {
+  const savedObjectsClient = getSavedObjectsClient(server);
+  savedObjectsClient.create('telemetry', fileUploadTelemetry, {
+    id: TELEMETRY_DOC_ID,
+    overwrite: true,
+  });
+}
+
+export function getSavedObjectsClient(server: Server): any {
+  const { SavedObjectsClient, getSavedObjectsRepository } = server.savedObjects;
+  const callWithInternalUser = callWithInternalUserFactory(server);
+  const internalRepository = getSavedObjectsRepository(callWithInternalUser);
+  return new SavedObjectsClient(internalRepository);
+}
+
+export async function incrementFileDataVisualizerIndexCreationCount(server: Server): Promise<void> {
```

Comment:
I'd recommend emulating what KQL's telemetry endpoint is doing here: https://github.com/elastic/kibana/blob/master/src/legacy/core_plugins/kibana/server/routes/api/kql_telemetry/index.js#L35

Reply:
Ok, planning to do another pass on telemetry. I'll give this a look!

Reply:
Sorry for the belated follow up! I diverted for some time to the related #34748 which focuses on the front-end portion. I needed this context to determine how well the plugin would work in practice for full round-trips. 

I did end up revising the telemetry to be more similar to what kql was doing but I'm certain it could use another pass. I think (hopefully) we're in a reasonable spot now but it's worth nothing that we'll end up iterating on telemetry at least one more round (and maybe more) for #35809.